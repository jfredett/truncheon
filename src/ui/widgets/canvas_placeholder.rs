use ratatui::{buffer::Buffer, layout::Rect, style::Color, widgets::{canvas::{Canvas, Rectangle}, Block, StatefulWidget, Widget}};



#[derive(Debug, Default, Clone)]
pub struct CanvasPlaceholder {
}

impl StatefulWidget for &CanvasPlaceholder {
    type State = ();

    fn render(self, area: Rect, buf: &mut Buffer, _state: &mut Self::State) {
        Widget::render(self, area, buf);
    }
}

impl Widget for &CanvasPlaceholder {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let widget = Canvas::default()
            .x_bounds([-100.0, 100.0])
            .y_bounds([-100.0, 100.0])
            .block(Block::bordered().title("Canvas"))
            .paint(|ctx| {
                ctx.draw(&Rectangle {
                    x: 10.0,
                    y: 20.0,
                    width: 10.0,
                    height: 10.0,
                    color: Color::Red,
                });
            });

        Widget::render(&widget, area, buf);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use rstest::rstest;

    mod rendering {

        use ratatui::style::Style;

        use super::*;

        #[rstest]
        fn renders_as_expected_stateless() {
            let rect = Rect::new(0, 0, 8, 8);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = CanvasPlaceholder::default();

            Widget::render(&placeholder, rect, &mut buffer);

            assert_debug_snapshot!(buffer);
        }

        #[rstest]
        fn renders_as_expected_stateful() {
            let rect = Rect::new(0, 0, 8, 8);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = CanvasPlaceholder::default();

            StatefulWidget::render(&placeholder, rect, &mut buffer, &mut ());

            assert_debug_snapshot!(buffer);
        }
    }
}

