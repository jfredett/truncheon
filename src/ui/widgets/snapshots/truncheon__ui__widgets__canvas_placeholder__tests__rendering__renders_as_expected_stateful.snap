---
source: src/ui/widgets/canvas_placeholder.rs
expression: buffer
---
Buffer {
    area: Rect { x: 0, y: 0, width: 8, height: 8 },
    content: [
        "┌Canvas┐",
        "│      │",
        "│      │",
        "│   ⠃  │",
        "│      │",
        "│      │",
        "│      │",
        "└──────┘",
    ],
    styles: [
        x: 0, y: 0, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 1, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 1, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 2, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 2, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 3, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 4, y: 3, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 5, y: 3, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 3, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 4, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 4, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 5, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 5, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 6, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 6, fg: White, bg: Black, underline: Reset, modifier: NONE,
    ]
}
